src/System.Text.Json/src/System/Text/Json/Serialization/CachedUtf8JsonWriter.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+
+namespace System.Text.Json.Serialization
+{
+    internal sealed class CachedUtf8JsonWriter
+    {
+        [ThreadStatic]
+        private static CachedUtf8JsonWriter s_cachedInstance;
+
+        private readonly Utf8JsonWriter _writer;
+
+#if DEBUG
+        private bool _inUse;
+#endif
+
+        private CachedUtf8JsonWriter(IBufferWriter<byte> stream, JsonWriterOptions writerOptions)
+        {
+            _writer = new Utf8JsonWriter(stream, writerOptions);
+        }
+
+        public static CachedUtf8JsonWriter Get(IBufferWriter<byte> stream, JsonWriterOptions writerOptions)
+        {
+            var writer = s_cachedInstance ?? new CachedUtf8JsonWriter(stream, writerOptions);
```

Comment:
I believe this check needs to include a comparison between the existing and new writer options. @steveharter, do we need an ==/equality operator on the writer options here to help with this check or would it make sense to just compare the indented bool? How would the caching be affected once we add more options like the `JavascriptEncoder`?

Reply:
Should I use `ConcurrentDictionary` to cache instances by options?

Reply:
> CanI use `ConcurrentDictionary` to cache instances by options? Then I need implement `IEquatable` for `JsonWriterOptions`, right?

Yes, that approach could work. However, thinking about this a bit more, we may want to step back a bit and see what we are gaining from caching the writer and if the additional complexity is worth it here.

One concern with caching multiple instances of the writer (by options) is that the writer ends up becoming long-lived. And also, given the default buffer size of 16k, the only saving here for many payloads would be the allocation of the writer. I am not sure if that's worth the trade-off here.

We should measure if caching makes a noticeable difference in performance before diving further into the implementation. I would try to see what type of benefit we are getting by running perf tests like these: https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/corefx/System.Text.Json/Serializer/WriteJson.cs

See the corefx workflow to see how to run these tests without/with your changes:
https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-corefx.md#preventing-regressions

>  Then I need implement `IEquatable` for `JsonWriterOptions`, right?

Regardless of whether we cache the writer within the serializer, I think we should make change the `JsonWriterOptions` to implement `IEquatable` anyway (it follows the .NET guidelines for structs). I can file an issue for this.

Reply:
I run existing benchmarks before my changes om empty master and after I implemented ``ConcurrentDictionary`
![Screenshot_1](https://user-images.githubusercontent.com/9221694/57575980-89410b80-745e-11e9-9552-ade4b4d37832.png)

Reply:
If I am reading this right, there doesn't seem to be much benefit from re-using the writer, in terms of throughput to justify the complexity. How's the allocation? Presumably, we save the cost of allocating the writer (which isn't much). I wonder if there is a benefit for writing really large JSON (say 100 KB), since the main benefit of re-using the writer is avoiding the resize/data copy within the underlying array.